corehq/apps/dhis2/api.py

```diff
@@ -0,0 +1,120 @@
+import httplib
+import json
+import logging
+
+import requests
+
+
+logger = logging.getLogger('json_api_logger')
+
+
+class JsonApiError(Exception):
+    """
+    JsonApiError is raised for HTTP or socket errors.
+    """
+    pass
+
+
+def json_serializer(obj):
+    """
+    A JSON serializer that serializes dates and times
+    """
+    if hasattr(obj, 'isoformat'):
+        return obj.isoformat()
+
+
+def log_request(func):
+
+    def request_wrapper(self, path, data=None, **kwargs):
+        log = logger.debug
+        request_error = ''
+        response_status = None
+        response_body = ''
+        try:
+            # This assumes that we always send data with post and put
+            status, response = func(self, path, **kwargs) if data is None else func(self, path, data, **kwargs)
+            response_status = status
+            response_body = '' if response is None else json.dumps(response)
+        except Exception as err:
+            log = logger.error
+            request_error = str(err)
+            raise err
+        else:
+            return status, response
+        finally:
+            log({
+                'domain': self.domain_name if self.domain_name is not None else '[N/A]',
+                'request_method': func.__name__.upper(),
+                'request_url': self.server_url + path,
+                'request_headers': json.dumps(self.headers),
+                'request_params': json.dumps(kwargs),
+                'request_body': '' if data is None else json.dumps(data, default=json_serializer),
+
+                'request_error': request_error,
+                'response_status': response_status,
+                'response_body': response_body,
+            })
+
+    return request_wrapper
+
+
+class JsonApiRequest(object):
+    """
+    Wrap requests with URL, header and authentication for DHIS2 API
+    """
+
+    def __init__(self, server_url, username, password, domain_name=None):
+        self.server_url = server_url  # e.g. "https://dhis2.example.com/api/26/"
+        self.headers = {'Accept': 'application/json'}
+        self.auth = (username, password)
+        self.domain_name = domain_name
+
+    @staticmethod
+    def json_or_error(response):
+        """
+        Return HTTP status, JSON
+
+        :raises JsonApiError: if HTTP status is not in the 200 (OK) range
+        """
+        if 200 <= response.status_code < 300:
+            return response.status_code, response.json() if response.content else None
```

Comment:
I don't see a huge advantage to not returning the full response object here. At least I think this should return a namedtuple.